static/templyte/src/app/download-template/download-template.component.ts

```diff
@@ -64,6 +74,50 @@ export class DownloadTemplateComponent implements OnInit {
         this.authService.logout(null);
     }
 
+    storeRating(new_rating) {
+        const component = this;
+        let authorUID:string  =  component.user.uid;
+        component.templateRatingsInfoDatabaseRef.once('value').then(snapshot2 => {
+            const old_rating = snapshot2.child(authorUID).val(); //value of previous rating
+            let hasRated:boolean = old_rating != null;
+            if (hasRated == true) {
+                // author has reviewed before
+                // update values in template directory
+                component.templateDirectoryInfoDatabaseRef.child('/ratingSum').once('value', function(snapshot) {
+                    component.templateDirectoryInfoRef.update({
+                        'ratingSum': snapshot.val() - old_rating
+                    })
+                });
+                component.templateDirectoryInfoDatabaseRef.child('/numberRatings').once('value', function(snapshot) {
+                    component.templateDirectoryInfoRef.update({
+                        'numberRatings': snapshot.val() - 1
+                    })
+                });
+            }
+            component.templateRatingsInfoRef.set({
+                [authorUID]: new_rating
+            });
+
+            //update all with new_rating
+            let ratingSum, ratingCount
+            component.templateDirectoryInfoDatabaseRef.child('/ratingSum').once('value', sumSnapshot => {
```

Comment:
Should use a transaction when possible (https://firebase.google.com/docs/firestore/manage-data/transactions).